### PR TITLE
Integrate cells into list type

### LINKED ISSUE #1085: Integrate cells into list type
**Description:**
$subject.

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject. 
Fixes #1085

### CODE REVIEW

File: compiler/modules/types/cell.bal [hunk 1]
```diff
@@ -128,10 +128,46 @@ function intersectCellAtomicType(CellAtomicType c1, CellAtomicType c2) returns C
...
+}
+
+function cellSubtypeComplement(SubtypeData t) returns SubtypeData {
+    SubtypeData subtypeData = bddSubtypeComplement(t);
+    return cellFixSubtypeData(subtypeData);
+}
+
+// SubtypeData being a boolean would result in a BasicTypeBitSet which could either be 0 or 1 << BT_CELL.
+// This is to avoid cell type being a BasicTypeBitSet, as we don't want to lose the cell wrapper.
+function cellFixSubtypeData(SubtypeData subtypeData) returns SubtypeData {
```
Comment (lochana-chathura) [inline compiler/modules/types/cell.bal:153]: Well, when we have two cells in which the diff is empty, we end up creating `0` as the resultant semType. This is contrary to having MemberSemType enforced with the following type definition.
``` bal 
public type MemberSemType readonly & record {|
    0 all = 0;
    BasicTypeBitSet some = CELL;
    ProperSubtypeData[1] subtypeDataList;
|};
```
Therefore I did this fix.

The option would be to make MemberSemType 
```bal
type MemberSemType NEVER | CELL | record {| 0 all = 0 BasicTypeBitSet some = CELL; ProperSubtypeData[1] subtypeDataList; |};
```
However, with the second option we will not have the cell wrapper all the time.

File: compiler/modules/types.regex/tests/data/regexInclusionTests.json [hunk 2]
```diff
@@ -17,8 +18,6 @@
     ["<", "(AC)*F", "(A(B)*C)*F"],
     ["=", "(AB*C)*F", "(A(B)*C)*F"],
 
-    ["<", "(RgeRgF(b((OS)*cDVB)*H)*kYLy)*W", "(RgeRgF(b(((glPrQ)*OS)*cDVB)*H)*kYLy)*W"],
```
Comment (lochana-chathura) [inline compiler/modules/types.regex/tests/data/regexInclusionTests.json:20]: https://github.com/ballerina-platform/nballerina/issues/1088#issuecomment-1276165054

File: compiler/modules/types/core.bal [hunk 3]
```diff
@@ -1008,22 +1012,36 @@ public function intSubtypeConstraints(SemType t) returns IntSubtypeConstraints?
...
 
-public function listAtomicSimpleArrayMemberType(ListAtomicType? atomic) returns BasicTypeBitSet? {
+public function defineListTypeWrapped(ListDefinition ld, Env env, SemType[] initial = [], int fixedLength = initial.length(), SemType rest = NEVER, CellMutability mut = CELL_MUT_LIMITED) returns SemType {
+    MemberSemType[] initialCells = from var i in initial select cellContaining(env, i, mut);
+    MemberSemType restCell = cellContaining(env, rest, mut);
+    return ld.define(env, initialCells, fixedLength, restCell);
+}
+
+public function tupleTypeWrapped(Env env, SemType... members) returns SemType {
+    ListDefinition def = new;
```
Comment (jclark) [inline compiler/modules/types/core.bal]: I would inline this and similar below.

File: compiler/modules/types/core.bal [hunk 4]
```diff
@@ -1008,22 +1012,36 @@ public function intSubtypeConstraints(SemType t) returns IntSubtypeConstraints?
     } 
 }
 
-public function listAtomicSimpleArrayMemberType(ListAtomicType? atomic) returns BasicTypeBitSet? {
+public function defineListTypeWrapped(ListDefinition ld, Env env, SemType[] initial = [], int fixedLength = initial.length(), SemType rest = NEVER, CellMutability mut = CELL_MUT_LIMITED) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/core.bal]: Wrapped/Deref isn't very symmmetric, but I don't have a better idea.

File: compiler/modules/types/mapping.bal [hunk 5]
```diff
@@ -1,6 +1,8 @@
 // Implementation specific to basic type list.
 
-public type Field [string, MemberSemType];
+public type Field [string, SemType];
```
Comment (jclark) [inline compiler/modules/types/mapping.bal:3]: Why do we need both Field and MemberField here?

**Review by jclark — COMMENTED** [review]
This looks very good generally.

We need to think a bit more about the extent to which List/Mapping know about cells.  We started off trying to keep these completely separate i.e. List/Mapping know nothing about cells.  But we have moved away from this, because of the need to maintain cell atomic types and because we can use readonly cells for function argument lists.

The current concept is that List/Mappings know that they have "member" semtypes but not that these semtypes are cells.  But this abstraction isn't fully maintained: list.bal accesses cells directly and has Deref functions.  It's not clear whether it is helpful to maintain distinction between MemberSemType and semtype that is a subtype of Cell.

I think we need to get clearer on this and implement a policy consistently.

Related to this is the Deref/Wrapped naming convention.

File: compiler/modules/types/mapping.bal (hunk 5 above)
Comment (lochana-chathura) [reply compiler/modules/types/mapping.bal:3]: Well, in the `defineListTypeWrapped()` we take a `SemType[]` and convert it to a `CellSemType[]`.
Similarly in the `defineMappingTypeWrapped()` we take a `Field[]` and convert it to a `CellField[]`. So we need two type definitions to represent `[string, SemType]`(input) and `[string, CellSemType]`(after wrapping) The current occurrence of `CellField` is only `mapping.bal`.

File: compiler/modules/types/core.bal (hunk 4 above)
Comment (lochana-chathura) [reply compiler/modules/types/core.bal]: replaced `deref` with `inner` with [1bddfa0](https://github.com/ballerina-platform/nballerina/pull/1091/commits/1bddfa0c4c379d8e38a70c5e7968a6f797dc461f) as per our discussion

**Pull request merged**